core/app.go

```diff
@@ -97,11 +98,42 @@ func getEnvVarNameFromService(service string) string {
 
 // Run starts the application and blocks until finished
 func (a *App) Run() {
-	a.handleSignals()
 	for {
+		ctx, cancel := context.WithCancel(context.Background())
+
+		// This provides an escape hatch for shutting down after all jobs have
+		// completed. Each time a job completes, during its cleanup func, it
+		// will be set `IsComplete` to `true`. Then it'll send `true` through
+		// this channel to wake up this goroutine, check all jobs for
+		// `IsComplete`, and shutdown ContainerPilot if all jobs are
+		// completed. This is because ContainerPilot is NOT a server and must
+		// shut down when no longer required (i.e. process containers).
+		completedCh := make(chan bool)
```

Comment:
This is the part I'm unhappy about but I don't believe there is any other way around it.

Every other case for using `context` throughout ContainerPilot allows the application to uniformly shutdown when calling `context.CancelFunc` which is passed into various parts of the application. This works great for the Reload endpoint or SIGTERM handling where `context` can be used by a single vector for termination.

This is not the case when handling shutdown when all Jobs have exited.

All Jobs running in ContainerPilot can cause a shutdown upon completion if no other jobs are running. This means we need to check the status of every job before calling `context.CancelFunc`. This goroutine provides that behavior for every Job that is exiting by providing a channel back to a thread which can properly exit ContainerPilot.